File: core/src/main/java/org/adoptopenjdk/jheappo/io/HeapProfile.java
package org.adoptopenjdk.jheappo.io;

import java.io.BufferedInputStream;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Path;


public final class HeapProfile {
    private final Path path;
    private final DataInputStream input;
    private boolean heapDumpEnd = false;

    public HeapProfile(Path path, DataInputStream input) {
        this.path = path;
        this.input = input;
    }

    public boolean isAtHeapDumpEnd() throws IOException {

        return heapDumpEnd || input.available() == 0;
    }

    public HeapProfileHeader readHeader() throws IOException {
        var header = new HeapProfileHeader();
        header.extract(input);

        return header;
    }

    private EncodedChunk readBody(DataInputStream inputStream, int bufferLength) throws IOException {
        byte[] buffer = new byte[bufferLength];
        int bytesRead = inputStream.read(buffer);

        if (bytesRead < bufferLength) {
            this.heapDumpEnd = true;

            throw new IOException("bytes request exceeded bytes read");
        } else {

            return new EncodedChunk(buffer);
        }
    }

    public HeapProfileRecord extract() throws IOException {
        byte tag = (new EncodedChunk(new byte[]{this.input.readByte()})).extractU1();
        long timeStamp = (long) this.input.readInt();
        int bodySize = this.input.readInt();

        final var body = readBody(input, bodySize);

        switch (tag) {
            case UTF8StringSegment.TAG:

                return new UTF8StringSegment(body);
            case LoadClass.TAG:

                return new LoadClass(body);
            case UnloadClass.TAG: {
                System.out.println("UnloadClass");

                return new UnloadClass(body);
            }
            case StackFrame.TAG:

                return new StackFrame(body);
            case StackTrace.TAG:

                return new StackTrace(body);
            case AllocSites.TAG: {
                System.out.println("AllocSites");

                return new AllocSites(body);
            }
            case HeapSummary.TAG: {
                System.out.println("HeapSummary");

                return new HeapSummary(body);
            }
            case StartThread.TAG: {
                System.out.println("StartThread");

                return new StartThread(body);
            }
            case HeapDumpSegment.TAG1:
            case HeapDumpSegment.TAG2:

                return new HeapDumpSegment(body);
            case HeapDumpEnd.TAG: {
                System.out.println("HeapDumpEnd");
                heapDumpEnd = true;

                return new HeapDumpEnd(body);
            }
            case CPUSamples.TAG: {
                System.out.println("CPUSamples");

                return new CPUSamples(body);
            }
            case ControlSettings.TAG: {
                System.out.println("ControlSettings");

                return new ControlSettings(body);
            }
            default:
                throw new IOException("Unknown record type " + tag);
        }

    }

    @Override
    public String toString() {

        return path.toString();
    }

    public static HeapProfile open(Path path, InputStream inputStream) {
        var input = new DataInputStream(new BufferedInputStream(inputStream));

        return new HeapProfile(path, input);
    }
}

File: parser/src/main/java/org/adoptopenjdk/jheappo/parser/HeapProfile.kt
package org.adoptopenjdk.jheappo.parser

/*
 * Copyright (c) 2018 Kirk Pepperdine.
 * Licensed under https://github.com/AdoptOpenJDK/jheappo/blob/master/LICENSE
 * Instructions: https://github.com/AdoptOpenJDK/jheappo/wiki
 */


import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.*
import java.nio.file.Path

class HeapProfile(private val path: Path, private val input: DataInputStream) {
    companion object {
        private val logger: Logger = LoggerFactory.getLogger(HeapProfile::class.java)

        fun open(path: Path, inputStream: InputStream): HeapProfile {
            val input = DataInputStream(BufferedInputStream(inputStream))
            return HeapProfile(path, input)
        }
    }

    private var heapDumpEnd: Boolean = false

    val isAtHeapDumpEnd: Boolean
        get() = heapDumpEnd || input.available() == 0


    fun readHeader(): HeapProfileHeader {
        val header = HeapProfileHeader()
        header.extract(input)
        return header
    }

    private fun readBody(inputStream: DataInputStream, bufferLength: Int): EncodedChunk {
        val buffer = ByteArray(bufferLength)
        // TODO allow short reads
        val bytesRead = inputStream.read(buffer)
        if (bytesRead < bufferLength) {
            heapDumpEnd = true
            throw IOException("bytes request exceeded bytes read")
        }
        return EncodedChunk(buffer)
    }

    fun extract(): HeapProfileRecord {
        val tag = EncodedChunk(byteArrayOf(input.readByte())).extractU1()
        val timeStamp = input.readInt().toLong()
        val bodySize = input.readInt()
        when (tag) {
            UTF8StringSegment.TAG -> return UTF8StringSegment(readBody(input, bodySize))
            LoadClass.TAG -> return LoadClass(readBody(input, bodySize))
            UnloadClass.TAG -> {
                logger.debug("UnloadClass")
                return UnloadClass(readBody(input, bodySize))
            }
            StackFrame.TAG -> return StackFrame(readBody(input, bodySize))
            StackTrace.TAG -> return StackTrace(readBody(input, bodySize))
            AllocSites.TAG -> {
                logger.debug("AllocSites")
                return AllocSites(readBody(input, bodySize))
            }
            HeapSummary.TAG -> {
                logger.debug("HeapSummary")
                return HeapSummary(readBody(input, bodySize))
            }
            StartThread.TAG -> {
                logger.debug("StartThread")
                return StartThread(readBody(input, bodySize))
            }
            EndThread.TAG -> {
                logger.debug("EndThread")
                return EndThread(readBody(input, bodySize))
            }
            HeapDumpSegment.TAG1, HeapDumpSegment.TAG2 -> return HeapDumpSegment(readBody(input, bodySize))
            HeapDumpEnd.TAG -> {
                logger.debug("HeapDumpEnd")
                heapDumpEnd = true
                return HeapDumpEnd(readBody(input, bodySize))
            }
            CPUSamples.TAG -> {
                logger.debug("CPUSamples")
                return CPUSamples(readBody(input, bodySize))
            }
            ControlSettings.TAG -> {
                logger.debug("ControlSettings")
                return ControlSettings(readBody(input, bodySize))
            }
            else -> throw IOException("Unknown record type + $tag")
        }
    }

    override fun toString(): String {
        return path.toString()
    }
}
